File: modules/dotnet/Morphir.CodeModel/projectSystem.fs
namespace Morphir.CodeModel.ProjectSystem

type ProjectName = string
type SourceDirectoryPath = string
type LocalDependency =
    | LocalDependency of string
type Dependency =
    | Dependency of string

type EntryPoint =
    | FQName of string
    //|FQNameUrl of string

type IRLocation = IRLocation of string
type StorageLocation = StorageLocation of string

type DecorationInfo =
    {
        DisplayName: string
        EntryPoint: EntryPoint
        IR: IRLocation
        StorageLocation: StorageLocation
    }

type MorphirProject = {
    Name: ProjectName
    SourceDirectory: SourceDirectoryPath
    LocalDependencies: LocalDependency list
    Dependencies: Dependency list
    Decorations: Map<string, DecorationInfo>
}

//TODO: Flesh out the MorphirManifest type
type MorphirManifest = MorphirManifest

File: modules/dotnet/Morphir.CodeModel/CodeModel/ProjectSystems/ProjectSystem.cs
using Dunet;
using Morphir.CodeModel.Common;
using Vogen;

namespace Morphir.CodeModel.ProjectSystems;

[ValueObject<string>]
public partial struct ProjectName
{
}

[ValueObject<string>]
public partial struct SourceDirectoryPath
{
}

public record DecorationInfo(
    DisplayName DisplayName,
    EntryPoint EntryPoint,
    IRLocation IR,
    StorageLocation StorageLocation);

public record MorphirProject(
    ProjectName Name,
    SourceDirectoryPath SourceDirectory,
    IReadOnlyList<LocalDependency> LocalDependencies,
    IReadOnlyList<Dependency> Dependencies);

[Union]
public partial record EntryPoint
{
    partial record ByFQName(string FQName);
}

[Union]
public partial record IRLocation
{
}

[Union]
public partial record StorageLocation
{
}

[Union]
public partial record LocalDependency
{
    partial record FromPath(string PathString);
}

[Union]
public partial record Dependency
{
    partial record FromPath(string PathString);
}